src/linker/Linker.Steps/MarkStep.cs

```diff
@@ -1380,18 +1380,42 @@ void MarkEntireAssembly (AssemblyDefinition assembly)
 			foreach (TypeDefinition type in module.Types)
 				MarkEntireType (type, new DependencyInfo (DependencyKind.TypeInAssembly, assembly));
 
-			foreach (ExportedType exportedType in module.ExportedTypes) {
-				MarkingHelpers.MarkExportedType (exportedType, module, new DependencyInfo (DependencyKind.ExportedType, assembly));
-				MarkingHelpers.MarkForwardedScope (new TypeReference (exportedType.Namespace, exportedType.Name, module, exportedType.Scope));
+			// Mark scopes of type references and exported types.
+			new TypeReferenceMarker (assembly, MarkingHelpers).Process ();
+		}
+
+		class TypeReferenceMarker : TypeReferenceWalker
+		{
+
+			readonly MarkingHelpers markingHelpers;
+
+			public TypeReferenceMarker (AssemblyDefinition assembly, MarkingHelpers markingHelpers)
+				: base (assembly)
+			{
+				this.markingHelpers = markingHelpers;
 			}
 
-			// Mark scopes of type references by traversing the assembly.
-			new TypeReferenceMarker (assembly, MarkingHelpers).Process ();
+			protected override void ProcessTypeReference (TypeReference type)
+			{
+				markingHelpers.MarkForwardedScope (type);
+			}
+
+			protected override void ProcessExportedType (ExportedType exportedType)
+			{
+				markingHelpers.MarkExportedType (exportedType, assembly.MainModule, new DependencyInfo (DependencyKind.ExportedType, assembly));
+				markingHelpers.MarkForwardedScope (new TypeReference (exportedType.Namespace, exportedType.Name, assembly.MainModule, exportedType.Scope));
+			}
 
-			// Also mark the scopes of metadata typeref rows to cover any not discovered by the traversal.
-			// This can happen when the compiler emits typerefs into IL which aren't strictly necessary per ECMA 335.
-			foreach (TypeReference typeReference in module.GetTypeReferences ())
-				MarkingHelpers.MarkForwardedScope (typeReference);
+			protected override void ProcessExtra ()
+			{
+				// Also mark the scopes of metadata typeref rows to cover any not discovered by the traversal.
+				// This can happen when the compiler emits typerefs into IL which aren't strictly necessary per ECMA 335.
+				foreach (TypeReference typeReference in assembly.MainModule.GetTypeReferences ()) {
```

Comment:
Side question (ignore for PR purposes): Would be interesting to know what would happen if we removed such type ref.  If it's not used, then it should not matter that we don't preserve the type forwarder for it. I guess I don't know what Cecil will do when saving the assembly (will it rebuild the type-ref table from scratch, I guess it has to), but then again that will not happen for "copy".

Reply:
Good point - I think there would be no harm in not marking these, at least for the final output (which will just have unused dangling typerefs). There could be issues like https://github.com/dotnet/linker/issues/2267 since the in-memory representation would be unresolvable, but arguably custom steps should never be looking at these typerefs anyway.